File: Python/Regression/50Startups-MultiLinearRegression.py
import pandas as pd 
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
from sklearn.linear_model import LinearRegression


# Import Dataset
df = pd.read_csv(r"C:\Users\snapp\Data-Projects\Datasets\50_Startups.csv")

# Independent Variables 
X = df.iloc[:,:-1]
# X = df.iloc[:,:-1].values
# Dependent Variables 
y = df.iloc[:,-1]

# Encoding Categorical Data 
onehotencoder = OneHotEncoder(handle_unknown='ignore')
enc_df = pd.DataFrame(onehotencoder.fit_transform(X[['State']]).toarray())
X = X[['R&D Spend', 'Administration', 'Marketing Spend']].join(enc_df)
X = X.iloc[:,:]

# Avoid Dummy Variable Trap (some Libraries like sikit learn.linearregression take care of dummy variables)
X = X.iloc[:, :-1]

# Split Data into Train and Test 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Fit Multiple Linear Regression to the Training Set 
MLR = LinearRegression()
MLR.fit(X_train, y_train)

# Predict test set 
y_pred = MLR.predict(X_test)

# Adjusted_R_squared = 1 – [(1-R2)*(n-1)/(n-k-1)] : n: number of samples - k: number of regressors
Adj_R2 = 1 - (1-MLR.score(X, y))*(len(y)-1)/(len(y)-X.shape[1]-1)

File: R/Regression/50_Startups_MultiLinearRegression.R
# Multiple Linear Regression 

# Import Data set 
df = read.csv("C:/Users/snapp/Data-Projects/Datasets/50_Startups.csv")

# Encode Categorical Data 
df$State = factor(df$State,
                  levels = c('New York','California','Florida'),
                  labels = c(1,2,3))

# Set random state for seeing same data 
set.seed(123)

# Split Dataset into train test split 
library(caTools)
split = sample.split(df$Profit, SplitRatio = 0.8)
train_set = subset(df, split == TRUE)
test_set = subset(df, split == FALSE)

# Feature Scaling: It is not necessary this will take care of with functions in model

# Fit multiple Linear Regression Model to Train set 
MLR = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
         data = train_set)
# You can choose all columns by . MLR = lm(formula = Profit ~ .,data = train_set)

summary(MLR)

# Predict the Test Set Result 
y_pred = predict(MLR, newdata = test_set)
y_pred






